types/lodash/index.d.ts

```diff
@@ -6366,98 +6366,34 @@ declare namespace _ {
         /**
          * @see _.forEach
          */
-        each<T>(
-            collection: T[],
-            iteratee?: ListIterator<T, any>
-        ): T[];
-
-        /**
-         * @see _.forEach
-         */
-        each(
-            collection: string,
+        each<TString extends string | null | undefined>(
+            collection: TString,
             iteratee?: ListIterator<string, any>
-        ): string;
+        ): TString;
 
         /**
          * @see _.forEach
          */
-        each<T>(
-            collection: List<T>,
+        each<T, TList extends List<T> | null | undefined>(
```

Comment:
I think I found a workaround:

```
forEach<T, TList extends List<T> | null | undefined>(
    collection: TList & (List<T> | null | undefined),
    iteratee?: ListIterator<T, any>
): TList;
```

Unfortunately my `TSource` tick above seems to have no effect - the type of `collection` is still inferred to be `List<T>` (not `T[]`). Also, without `TSource`, you can still manually specify the type of `collection` to be a more specific type. Example:

```
_.each([1, 2], (value, index, collection) => {
    // collection is inferred to be List<number>,
    // regardless of whether I use the TSource trick or not
});
_.each([1, 2], (value, index, collection: number[]) => {
    // compiles successfully, regardless of whether I use the TSource trick or not
});
```

So I think I'll leave the `TSource` parameter out for simplicity.